### PR TITLE
New: pass cwd from cli engine

### PR DESCRIPTION
**Pull request by fa93hws**
**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

**What changes did you make? (Give an overview)**
Pass the cwd from the cli engine
https://github.com/eslint/eslint/issues/11218
RFC link: https://github.com/eslint/rfcs/pull/35

**Is there anything you'd like reviewers to focus on?**
`Linter.js`

### CODE REVIEW

File: lib/linter/linter.js [hunk 1]
```diff
@@ -1001,7 +1003,11 @@ const internalSlotsMap = new WeakMap();
  */
 class Linter {
 
-    constructor() {
+    /**
+     * @param {Object} config the config object
+     * @param {string | undefined} config.cwd cwd, can be undefined.
+     */
+    constructor({ cwd }) {
```
Comment (fa93hws) [inline lib/linter/linter.js]: Should I make the `config` as `Object | undefined` for the sake of backward compatibility?

Comment (kaicataldo) [reply lib/linter/linter.js]: Can we make this a non-breaking change by making the config object optional? e.g. `constructor({ cwd } = {}) {}`

File: lib/linter/linter.js [hunk 2]
```diff
@@ -1001,7 +1003,11 @@ const internalSlotsMap = new WeakMap();
  */
 class Linter {
 
-    constructor() {
+    /**
+     * @param {Object} config the config object
+     * @param {string | undefined} config.cwd cwd, can be undefined.
```
Comment (kaicataldo) [inline lib/linter/linter.js]: Can we describe what the `cwd` config option actually does?

File: lib/linter/linter.js [hunk 3]
```diff
@@ -1010,6 +1016,27 @@ class Linter {
         });
 
         this.version = pkg.version;
+        this._setCwd(cwd);
+    }
+
+    /**
+     * set this.cwd
```
Comment (kaicataldo) [inline lib/linter/linter.js]: It would be great if this JSDoc could be more descriptive.

File: lib/linter/linter.js (hunk 1 above)
Comment (mysticatea) [reply lib/linter/linter.js]: Yes, we should not make the breaking change.

File: lib/linter/linter.js [hunk 4]
```diff
@@ -1010,6 +1016,27 @@ class Linter {
         });
 
         this.version = pkg.version;
+        this._setCwd(cwd);
```
Comment (mysticatea) [inline lib/linter/linter.js]: Would you use `internalSlotsMap` to store private data? We have experienced that people use private data directly via `_foo`-like properties. Therefore we want to hide private data in public classes.

I'd like to recommend just putting `cwd` to the internal slots instead of `_setCwd`. For example,

```diff
        internalSlotsMap.set(this, {
+           cwd: normalizeCwd(cwd),
            lastConfigArray: null,
            lastSourceCode: null,
            parserMap: new Map([["espree", espree]]),
            ruleMap: new Rules()
        });
```

File: lib/linter/linter.js [hunk 5]
```diff
@@ -1010,6 +1016,27 @@ class Linter {
...
+    /**
+     * set this.cwd
+     * @param {string | undefined} cwd raw cwd
+     * @returns {void}
+     * @private
+     */
+    _setCwd(cwd) {
+        if (cwd) {
+            this._cwd = cwd;
+        } else if (process) {
```
Comment (mysticatea) [inline lib/linter/linter.js]: This needs `typeof` operator because of avoiding `ReferenceError`.

File: docs/developer-guide/nodejs-api.md [hunk 6]
```diff
@@ -109,7 +118,7 @@ You can call `verify()` like this:
 
 ```js
 const Linter = require("eslint").Linter;
-const linter = new Linter();
+const linter = new Linter({});
```
Comment (mysticatea) [inline docs/developer-guide/nodejs-api.md]: Would you remove unnecessary change such as this line? Small diff is great thing.

File: lib/linter/linter.js [hunk 7]
```diff
@@ -1134,7 +1159,8 @@ class Linter {
                 parserName,
                 settings,
                 options.filename,
-                options.disableFixes
+                options.disableFixes,
+                internalSlotsMap.get(this).cwd
```
Comment (mysticatea) [inline lib/linter/linter.js]:
```suggestion
                slots.cwd
```

File: lib/linter/linter.js [hunk 8]
```diff
@@ -1001,8 +1021,13 @@ const internalSlotsMap = new WeakMap();
  */
 class Linter {
 
-    constructor() {
+    /**
+     * @param {Object | undefined} config the config object
+     * @param {string | undefined} config.cwd  path to a directory that should be considered as the current working directory, can be undefined.
```
Comment (mysticatea) [inline lib/linter/linter.js]: This notation means that it requires one argument ant it can be an object or undefined. To represent optional arguments we use `[foo]`-like notation.

```
     * @param {Object} [config] the config object
     * @param {string} [config.cwd] path to a directory that should be considered as the current working directory, can be undefined.
```

Off-topic: Suggestion functionality doesn't seem to support multiline.

File: docs/developer-guide/nodejs-api.md (hunk 6 above)
Comment (fa93hws) [reply docs/developer-guide/nodejs-api.md]: No problem!
Do you have any plan to remove the optional config object in the constructor?
Personally I don't think optional parameter is a good practice.

File: docs/developer-guide/nodejs-api.md [hunk 9]
```diff
@@ -80,13 +80,22 @@ const codeLines = SourceCode.splitLines(code);
 
 ## Linter
 
-The `Linter` object does the actual evaluation of the JavaScript code. It doesn't do any filesystem operations, it simply parses and reports on the code. In particular, the `Linter` object does not process configuration objects or files. You can retrieve instances of `Linter` like this:
+The `Linter` object does the actual evaluation of the JavaScript code. It doesn't do any filesystem operations, it simply parses and reports on the code. In particular, the `Linter` object does not process configuration objects or files.
+The `Linter` is a constructor, and you can create a new instance by passing in the options you want to use. The available options are:
+
+* `cwd` - Path to a directory that should be considered as the current working directory and can be retrieved by calling `context.getCwd()` from the rule. (If the `cwd` is `undefined`, it will be normalized to `process.cwd()` if the `process` exists, or `undefined` otherwise)
```
Comment (kaicataldo) [inline docs/developer-guide/nodejs-api.md]: Wording suggestion:

```suggestion
* `cwd` - Path to a directory that should be considered as the current working directory. It is accessible to rules by calling `context.getCwd()`. If `cwd` is `undefined`, it will be normalized to `process.cwd()` if the global `process` object is defined (for example, in the Node.js runtime) , or `undefined` otherwise.
```

File: docs/developer-guide/nodejs-api.md [hunk 10]
```diff
@@ -80,13 +80,22 @@ const codeLines = SourceCode.splitLines(code);
...
+For example:
 
 ```js
 const Linter = require("eslint").Linter;
-const linter = new Linter();
+const linter1 = new Linter({ cwd: 'path/to/project' });
+const linter2 = new Linter();
 ```
 
+in this example, rules run on `linter1` will get `path/to/project` when calling `context.getCwd()`.
```
Comment (kaicataldo) [inline docs/developer-guide/nodejs-api.md]:
```suggestion
In this example, rules run on `linter1` will get `path/to/project` when calling `context.getCwd()`.
```

File: docs/developer-guide/nodejs-api.md [hunk 11]
```diff
@@ -80,13 +80,22 @@ const codeLines = SourceCode.splitLines(code);
...
+Those run on `linter2` would get `process.cwd()` if the `process` exists (e.g. under node environment) or `undefined` otherwise (e.g. on the browser https://eslint.org/demo).
```
Comment (kaicataldo) [inline docs/developer-guide/nodejs-api.md]: For consistency with the the previous sentence.
```suggestion
Those run on `linter2` will get `process.cwd()` if the global `process` object is defined or `undefined` otherwise (e.g. on the browser https://eslint.org/demo).
```

File: docs/developer-guide/working-with-rules.md [hunk 12]
```diff
@@ -126,6 +126,7 @@ The `context` object contains additional functionality that is helpful for rules
 Additionally, the `context` object has the following methods:
 
 * `getAncestors()` - returns an array of the ancestors of the currently-traversed node, starting at the root of the AST and continuing through the direct parent of the current node. This array does not include the currently-traversed node itself.
+* `getCwd()` - returns the `cwd` passed to [Linter](./nodejs-api.md#Linter), path to a directory that should be considered as the current working directory.
```
Comment (kaicataldo) [inline docs/developer-guide/working-with-rules.md]:
```suggestion
* `getCwd()` - returns the `cwd` passed to [Linter](./nodejs-api.md#Linter). It is a path to a directory that should be considered as the current working directory.
```

File: tests/lib/linter/linter.js [hunk 13]
```diff
@@ -3195,6 +3195,58 @@ describe("Linter", () => {
         });
     });
 
+    describe("when receiving cwd in options in construction", () => {
```
Comment (kaicataldo) [inline tests/lib/linter/linter.js]:
```suggestion
    describe("when receiving cwd in options during instantiation", () => {
```

File: docs/developer-guide/nodejs-api.md (hunk 9 above)
Comment (kaicataldo) [inline docs/developer-guide/nodejs-api.md]: Can we link to the `context.getCwd()` docs here?

**Comment by kaicataldo** [comment]
Ah, also forget to mention: can you change the commit message from `Chore:` to `Enhancement:`? Since this adds a new API, it should be a semver-minor change.

File: docs/developer-guide/nodejs-api.md (hunk 6 above)
Comment (mysticatea) [reply docs/developer-guide/nodejs-api.md]: No, I don't have. Because `options` is an option, people can omit it. What are the advantages to disallow it?

**Comment by mysticatea** [comment]
@kaicataldo I guess you meant `Update:` or `New:`.

File: docs/developer-guide/nodejs-api.md (hunk 6 above)
Comment (fa93hws) [reply docs/developer-guide/nodejs-api.md]: Personally I prefer to limit the type as much as possible.
If something is `A | undefined` then it need to be checked whether it's `A` or `undefined` for the entire scope.
It would be a heavy burden on mind especially in javascript because there is no type check.

Comment (mysticatea) [reply docs/developer-guide/nodejs-api.md]: For such a case, we have the default parameter syntax :)

For example, in [lib/cli-engine/file-enumerator.js#L190-L197](https://github.com/eslint/eslint/blob/master/lib/cli-engine/file-enumerator.js#L190-L197) case, users can omit options because of options, but every option isn't `undefined` in this scope explicitly.

**Comment by kaicataldo** [comment]
Thanks for contributing!
